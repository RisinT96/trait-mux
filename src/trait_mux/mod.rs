pub mod codegen;
pub mod lower;
pub mod parse;
