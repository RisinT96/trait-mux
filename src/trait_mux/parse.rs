//! This module provides functionality to parse a named list of traits or paths from a `TokenStream`.
//! It supports both simple trait names (e.g., `Display`) and full paths (e.g., `std::fmt::Display`).
//! The parsed traits are stored as `Path` objects in the `Ast` struct, along with the name of the implementation.

use proc_macro2::TokenStream;
use proc_macro_error::abort;
use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;
use syn::token::Comma;
use syn::{parse2, Ident, Path, Result, Token};

/// Represents the parsed Abstract Syntax Tree (AST) for a named list of traits or paths.
///
/// The syntax format is `SomeName{Display, std::fmt::Debug}`, where:
/// - `SomeName` is the name of the implementation.
/// - `{Display, std::fmt::Debug}` is a comma-separated list of traits or paths.
pub struct Ast {
    /// The name of the implementation (e.g., `SomeName`).
    pub name: Ident,
    /// A punctuated list of parsed paths representing traits or modules.
    /// Each path can be a simple identifier (e.g., `Display`) or a full path (e.g., `std::fmt::Display`).
    pub paths: Punctuated<Path, Comma>,
}

impl Parse for Ast {
    /// Parses a syntax like `SomeName{Display, std::fmt::Debug}`.
    ///
    /// # Arguments
    /// * `input` - The input stream to parse.
    ///
    /// # Returns
    /// * `Result<Self>` - The parsed `Ast` containing the identifier name and list of paths.
    ///
    /// # Errors
    /// Returns an error if the input does not match the expected syntax.
    fn parse(input: ParseStream) -> Result<Self> {
        let name = input.parse::<Ident>()?;

        let content;
        syn::braced!(content in input);

        let paths = Punctuated::<Path, Token![,]>::parse_terminated(&content)?;

        Ok(Ast { name, paths })
    }
}

/// Parses a `TokenStream` into an `Ast` containing a named list of paths.
///
/// The input must follow the syntax `SomeName{Display, std::fmt::Debug}`.
///
/// # Arguments
/// * `ts` - The `TokenStream` to parse.
///
/// # Returns
/// * `Ast` - The parsed AST.
///
/// # Panics
/// Panics if the input cannot be parsed, using the `abort!` macro to provide an error message.
pub fn parse(ts: TokenStream) -> Ast {
    match parse2::<Ast>(ts) {
        Ok(ast) => ast,
        Err(e) => {
            abort!(e.span(), e)
        }
    }
}

#[cfg(test)]
mod tests {
    //! Unit tests for the parsing functionality.
    //!
    //! These tests verify that the parser correctly handles valid and invalid inputs,
    //! including simple trait names, full paths, and edge cases.

    use super::*;
    use quote::quote;

    /// Tests parsing with the new syntax format: Name{traits...}.
    ///
    /// Verifies that the parser correctly extracts the name and paths.
    #[test]
    fn valid_named_syntax() {
        let ast = parse(quote!(SomeName{Display, std::fmt::Debug}));

        assert_eq!(ast.name.to_string(), "SomeName");
        assert_eq!(ast.paths.len(), 2);
        assert_eq!(ast.paths[0].get_ident().unwrap().to_string(), "Display");

        let debug = &ast.paths[1].segments;
        assert_eq!(debug.len(), 3);
        assert_eq!(debug[0].ident.to_string(), "std");
        assert_eq!(debug[1].ident.to_string(), "fmt");
        assert_eq!(debug[2].ident.to_string(), "Debug");
    }

    /// Tests parsing with mixed path formats.
    ///
    /// Verifies that the parser handles a mix of full paths and simple trait names.
    #[test]
    fn valid_syntax_mixed_paths() {
        let ast = parse(quote!(MyImpl{std::fmt::Display, ::fmt::Debug, Clone}));

        assert_eq!(ast.name.to_string(), "MyImpl");
        assert_eq!(ast.paths.len(), 3);

        // Check the segments of the path for the first trait
        let display = &ast.paths[0].segments;
        assert_eq!(display.len(), 3);
        assert_eq!(display[0].ident.to_string(), "std");
        assert_eq!(display[1].ident.to_string(), "fmt");
        assert_eq!(display[2].ident.to_string(), "Display");

        let debug = &ast.paths[1].segments;
        assert_eq!(debug.len(), 2);
        assert_eq!(debug[0].ident.to_string(), "fmt");
        assert_eq!(debug[1].ident.to_string(), "Debug");

        let clone = &ast.paths[2].segments;
        assert_eq!(clone.len(), 1);
        assert_eq!(clone[0].ident.to_string(), "Clone");
    }

    /// Tests parsing an empty list of traits with a name.
    ///
    /// Verifies that the parser correctly handles an empty list of traits.
    #[test]
    fn empty_named_trait_list() {
        let ast = parse(quote!(EmptyImpl {}));
        assert_eq!(ast.name.to_string(), "EmptyImpl");
        assert_eq!(ast.paths.len(), 0);
    }

    /// Tests parsing invalid input where a number is used instead of a valid path.
    ///
    /// Verifies that the parser fails when encountering invalid paths.
    #[test]
    #[should_panic]
    fn invalid_trait_input_not_path() {
        // Using a number instead of an identifier, which should cause the parser to fail
        parse(quote!(Invalid{Display, 123, Debug}));
        // The test should panic due to the abort! macro being called
    }

    /// Tests parsing invalid input with missing braces.
    ///
    /// Verifies that the parser fails when braces are missing.
    #[test]
    #[should_panic]
    fn invalid_trait_input_missing_braces() {
        parse(quote!(NoImpl));
    }

    /// Tests parsing invalid input with empty input.
    ///
    /// Verifies that the parser fails when the input is empty.
    #[test]
    #[should_panic]
    fn invalid_empty_input() {
        parse(quote!());
    }
}
