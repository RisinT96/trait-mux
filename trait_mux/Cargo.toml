[package]
authors.workspace = true
description.workspace = true
edition.workspace = true
keywords.workspace = true
license.workspace = true
name = "trait_mux"
readme.workspace = true
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
trait_mux_macros = { path = "../trait_mux_macros", version = "0.1.0" }

[dev-dependencies]
trybuild = "~1.0"

[[test]]
name = "integration"
path = "tests/ui.rs"
