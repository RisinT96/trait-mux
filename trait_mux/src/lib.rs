pub use trait_mux_macros::trait_mux;
