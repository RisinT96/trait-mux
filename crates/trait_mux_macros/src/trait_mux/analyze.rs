//! Analysis module that transforms the parsed AST into a Model.
//! This model represents all the information necessary to generate the output code.
//! The main responsibility is to extract traits and generate all possible enum variants
//! that will be used in the final generated code.

use proc_macro2::{Ident, Span};
use syn::{Path, spanned::Spanned};

use crate::parse::Ast;

/// The core model structure that contains all processed information from the AST.
/// This model is used as input for code generation, representing enum variants and traits
/// in a format that's easy to work with.
pub struct Model<'t> {
    /// The identifier of the main enum, taken from the AST.
    pub enum_ident: &'t Ident,
    /// All possible variants of the enum based on trait combinations.
    pub enum_variants: Vec<EnumVariant<'t>>,
    /// The identifier for the wrapper structure that will encapsulate the enum.
    pub wrap_ident: Ident,
    /// All traits extracted from the AST.
    pub traits: Vec<Trait<'t>>,
}

/// Represents a trait with its identifier and path.
/// Used to track traits throughout the code generation process.
#[derive(Copy, Clone)]
pub struct Trait<'t> {
    /// The identifier of the trait (the name).
    pub ident: &'t Ident,
    /// The full path to the trait, including any module qualifiers.
    pub path: &'t Path,
}

/// Represents an enum variant, including its identifier, and the traits it implements.
/// Each variant corresponds to a specific combination of implemented traits.
pub struct EnumVariant<'t> {
    /// The enum ident, used both for the enum variant name, and (if needed) for the trait aggregate
    /// name.
    /// e.g. `TypeBinaryDebugDisplay`
    pub ident: Ident,
    /// The traits that this enum implements.
    /// This is a subset of all traits defined in the Model.
    pub implemented_traits: Vec<Trait<'t>>,
}

/// Analyzes the AST and constructs a Model containing all the necessary information
/// for code generation. This is the main entry point for the analysis phase.
///
/// # Arguments
///
/// * `ast` - The parsed AST containing trait paths and enum name
///
/// # Returns
///
/// A Model containing the processed information ready for code generation
pub fn analyze(ast: &Ast) -> Model {
    let traits = extract_traits(ast);
    let enum_variants = generate_enum_variants(ast, &traits);
    let wrap_ident = Ident::new(&format!("Wrap{}", ast.name), Span::call_site());

    Model {
        enum_ident: &ast.name,
        enum_variants,
        wrap_ident,
        traits,
    }
}

/// Extracts traits from the given AST and converts them to the Trait model.
/// Emits an error if a path is empty or malformed.
///
/// # Arguments
///
/// * `ast` - The AST to extract traits from
///
/// # Returns
///
/// A vector of Trait structs sorted alphabetically by their identifiers
fn extract_traits(ast: &Ast) -> Vec<Trait> {
    let mut traits = vec![];

    for path in &ast.paths {
        if path.segments.is_empty() {
            proc_macro_error::emit_error!(
                path.span(),
                "unexpected end of input, expected identifier"
            );
        }

        // Unwrap safety: checked that segments is not empty.
        traits.push(Trait {
            ident: &path.segments.last().unwrap().ident,
            path,
        });
    }

    // Sort traits alphabetically by their identifier.
    traits.sort_by_key(|t| t.ident.to_string());

    traits
}

/// Generates all possible enum variants from the given traits.
/// The variants are sorted by descending length and then alphabetically.
/// The order is very important for later stages, as we want to generate code
/// with the most specific trait constraints first, and relax the constraints as
/// we go down, if the order was incorrect, autoref specialization won't work
/// properly.
///
/// # Arguments
///
/// * `ast` - The AST containing the enum name
/// * `traits` - A vector of Trait structs to generate permutations from
///
/// # Returns
///
/// A vector of EnumVariant structs representing all possible trait combinations
fn generate_enum_variants<'t>(ast: &Ast, traits: &Vec<Trait<'t>>) -> Vec<EnumVariant<'t>> {
    let mut permutations = Vec::new();
    let n = traits.len();

    // Create all possible permutations of the trait names.
    // We have 2^n possible permutations.
    for i in 0..(1 << n) {
        let mut permutation = vec![];

        for (j, r#trait) in traits.iter().enumerate() {
            if (i & (1 << j)) != 0 {
                permutation.push(*r#trait);
            }
        }
        permutations.push(permutation);
    }

    let mut variants = permutations
        .iter()
        .map(|variant| {
            let variant_name = if variant.is_empty() {
                "None".to_string()
            } else {
                variant
                    .iter()
                    .map(|t| t.ident.to_string())
                    .collect::<String>()
            };

            let variant_name = format!("{}{}", ast.name, variant_name);

            EnumVariant {
                ident: Ident::new(&variant_name, Span::call_site()),
                implemented_traits: variant.to_vec(),
            }
        })
        .collect::<Vec<_>>();

    // Sort by the length of implemented traits, then alphabetically.
    variants.sort_by_key(|e| format!("{} {}", e.implemented_traits.len(), e.ident));
    variants.reverse();

    variants
}
