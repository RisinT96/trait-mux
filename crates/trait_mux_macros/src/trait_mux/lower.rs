use convert_case::{Case, Casing};

use proc_macro2::{Ident, Span};
use syn::{Path, spanned::Spanned};

use crate::parse::Ast;

/// Intermediate representation (IR) of the parsed AST.
/// Contains the identifier, enum variants, and functions derived from the AST.
pub struct Ir<'t> {
    pub enum_ident: &'t Ident,
    pub enum_variants: Vec<EnumVariant<'t>>,
    pub enum_impl_functions: Vec<Function<'t>>,
}

/// Represents a trait with its identifier and path.
#[derive(Copy, Clone)]
pub struct Trait<'t> {
    pub ident: &'t Ident,
    pub path: &'t Path,
}

/// Represents an enum variant, including its identifier,
/// match identifier, and the traits it implements.
pub struct EnumVariant<'t> {
    pub ident: Ident,
    pub implemented_traits: Vec<Trait<'t>>,
}

/// Represents a function derived from a trait, including its identifier,
/// result path, and the variants it applies to.
pub struct Function<'t> {
    pub enum_variant: Ident,
    pub result_path: &'t Path,
    pub variants: Vec<Ident>,
}

/// Extracts traits from the given AST.
/// Emits an error if a path is empty or malformed.
fn extract_traits(ast: &Ast) -> Vec<Trait> {
    let mut traits = vec![];

    for path in &ast.paths {
        if path.segments.is_empty() {
            proc_macro_error::emit_error!(
                path.span(),
                "unexpected end of input, expected identifier"
            );
        }

        // Unwrap safety: checked that segments is not empty.
        traits.push(Trait {
            ident: &path.segments.last().unwrap().ident,
            path,
        });
    }

    // Sort traits alphabetically by their identifier.
    traits.sort_by_key(|t| t.ident.to_string());
    traits
}

/// Generates all possible enum variants from the given traits.
/// The variants are sorted by descending length and then alphabetically.
fn generate_variants<'t>(traits: &Vec<Trait<'t>>) -> Vec<EnumVariant<'t>> {
    let mut permutations = Vec::new();
    let n = traits.len();

    // Create all possible permutations of the trait names.
    // We have 2^n possible permutations.
    for i in 0..(1 << n) {
        let mut permutation = vec![];

        for (j, r#trait) in traits.iter().enumerate() {
            if (i & (1 << j)) != 0 {
                permutation.push(*r#trait);
            }
        }
        permutations.push(permutation);
    }

    let mut variants = permutations
        .iter()
        .map(|variant| {
            let variant_name = if variant.is_empty() {
                "None".to_string()
            } else {
                variant
                    .iter()
                    .map(|t| t.ident.to_string())
                    .collect::<String>()
            };

            EnumVariant {
                ident: Ident::new(&variant_name, Span::call_site()),
                implemented_traits: variant.to_vec(),
            }
        })
        .collect::<Vec<_>>();

    // Sort by the length of implemented traits, then alphabetically.
    variants.sort_by_key(|e| format!("{} {}", e.implemented_traits.len(), e.ident));
    variants.reverse();

    variants
}

/// Generates functions for each trait, mapping them to the enum variants
/// that implement the trait.
pub fn generate_enum_impl_functions<'t, 'p>(
    traits: &'p [Trait<'t>],
    variants: &'p [EnumVariant<'t>],
) -> Vec<Function<'t>> {
    traits
        .iter()
        .map(|t| {
            let name = format!("try_as_{}", t.ident.to_string().to_case(Case::Snake));

            // Find all enum variants that implement the trait `t`.
            let variants = variants
                .iter()
                .filter(|p| {
                    p.implemented_traits
                        .iter()
                        .any(|it| core::ptr::eq(it.path, t.path))
                })
                .map(|p| p.ident.clone())
                .collect();

            Function {
                enum_variant: Ident::new(&name, Span::call_site()),
                result_path: t.path,
                variants,
            }
        })
        .collect()
}

/// Converts the given AST into its intermediate representation (IR).
/// This includes extracting traits, generating enum variants, and creating functions.
pub fn lower(ast: &Ast) -> Ir {
    let traits = extract_traits(ast);
    let variants = generate_variants(&traits);
    let enum_impl_functions = generate_enum_impl_functions(&traits, &variants);

    Ir {
        enum_ident: &ast.name,
        enum_variants: variants,
        enum_impl_functions,
    }
}
