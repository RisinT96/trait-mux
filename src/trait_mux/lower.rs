use std::collections::HashSet;

use syn::spanned::Spanned;

use crate::parse::Ast;

pub struct Ir {
    pub enum_name: String,
    pub trait_names: Vec<String>,
    pub permutations: Vec<HashSet<String>>,
}

fn extract_trait_names(ast: &Ast) -> Vec<String> {
    let mut trait_names = vec![];

    for path in &ast.paths {
        if path.segments.is_empty() {
            proc_macro_error::emit_error!(
                path.span(),
                "unexpected end of input, expected identifier"
            );
        }

        // unwrap safety: checked that segments is not empty.
        trait_names.push(path.segments.last().unwrap().ident.to_string());
    }

    trait_names.sort();
    trait_names
}

fn generate_permutations(trait_names: &Vec<String>) -> Vec<HashSet<String>> {
    let mut permutations = Vec::new();
    let n = trait_names.len();

    // Create all possible permutations of the trait names.
    // We have 2^n possible permutations
    for i in 0..(1 << n) {
        let mut permutation = HashSet::new();

        for (j, name) in trait_names.iter().enumerate() {
            if (i & (1 << j)) != 0 {
                permutation.insert(name.clone());
            }
        }
        permutations.push(permutation);
    }

    // We want the list to be sorted descending by amount of traits, this is important as we'll be
    // using `spez` to generate specialization, where the order is very important, as the first
    // match (even not full match) will be the one that dictates the output.
    permutations.sort_by_key(|e| e.len());
    permutations.reverse();

    permutations
}

pub fn lower(ast: Ast) -> Ir {
    let trait_names = extract_trait_names(&ast);
    let permutations = generate_permutations(&trait_names);

    Ir {
        enum_name: ast.name.to_string(),
        trait_names,
        permutations,
    }
}
