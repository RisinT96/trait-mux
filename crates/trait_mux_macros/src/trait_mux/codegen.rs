use proc_macro::Ident;
use proc_macro2::{Span, TokenStream};
use quote::quote;

use crate::lower::Ir;

use super::lower::{AutorefSpecializer, Constraint, EnumVariant, Function, TraitAggregate};

fn refs(n: usize) -> TokenStream {
    let mut refs = TokenStream::new();
    for _ in 0..n {
        refs.extend(quote![&]);
    }
    refs
}

pub fn codegen(ir: Ir) -> TokenStream {
    let mut result = TokenStream::new();

    result.extend(generate_wrap(&ir));
    result.extend(generate_trait_aggregates(&ir));
    result.extend(generate_enum(&ir));
    result.extend(generate_enum_impl(&ir));
    result.extend(generate_autoref_specializers(&ir));

    let into = &ir.into;
    let into_tag = &ir.into_tag;
    let refs = refs(ir.wrap_derefs);
    let wrap = &ir.wrap_ident;

    result.extend(quote! {
        macro_rules! #into {
            ($var:tt) => {
                (#refs #wrap(&$var)).#into_tag().#into(&$var)
            }
        }
    });

    result
}

fn generate_wrap(ir: &Ir) -> TokenStream {
    let wrap = ir.wrap_ident;

    quote! {
        pub struct #wrap<'t, T>(pub &'t T);
    }
}

fn generate_trait_aggregates(ir: &Ir) -> TokenStream {
    let mut trait_aggregates = TokenStream::new();

    ir.trait_aggregates
        .iter()
        .map(|TraitAggregate { name, traits }| {
            let traits: Vec<_> = traits.iter().map(|t| t.path).collect();

            trait_aggregates.extend(quote! {
                pub trait #name: #(#traits)+* {}
                impl<T: #(#traits)+*> #name for T {}
            });
        })
        .count();

    trait_aggregates
}

fn generate_enum(ir: &Ir) -> TokenStream {
    let enum_name = ir.r#enum.name;

    let mut enum_fields = TokenStream::new();

    for EnumVariant { ident, constraint } in &ir.r#enum.variants {
        let constraint = match constraint {
            Constraint::None => quote! {},
            Constraint::Path(path) => quote! {(&'t dyn #path)},
            Constraint::Ident(ident) => quote! {(&'t dyn #ident)},
        };

        enum_fields.extend(quote! {
            #ident #constraint,
        });
    }

    quote! {
        pub enum #enum_name<'t> {
            #enum_fields
        }
    }
}

fn generate_enum_impl(ir: &Ir) -> TokenStream {
    let enum_name = ir.r#enum.name;

    let mut fns = TokenStream::new();

    for Function {
        name,
        result_path,
        matching_variants,
    } in &ir.enum_impl.functions
    {
        fns.extend(quote! {
            pub fn #name(&self) -> ::core::option::Option<&dyn #result_path> {
                match self {
                    #(#enum_name::#matching_variants (v) => Some(v),)*
                    _ => None,
                }
            }
        });
    }

    quote! {
        impl<'t> #enum_name<'t> {
            #fns
        }
    }
}

fn generate_autoref_specializers(ir: &Ir) -> TokenStream {
    let mut autoref_specializers = TokenStream::new();

    let enum_name = ir.r#enum.name;
    let wrap = ir.wrap_ident;
    let into = &ir.into;
    let into_tag = &ir.into_tag;

    ir.autoref_specializers
        .iter()
        .map(
            |AutorefSpecializer {
                 tag,
                 r#match,
                 deref_count,
                 variant,
                 constraint,
             }| {
                let refs = refs(*deref_count);

                let t_constraint = match constraint {
                    Constraint::None => quote! {},
                    Constraint::Path(path) => quote! {: #path},
                    Constraint::Ident(ident) => quote! {: #ident},
                };

                let param = match constraint {
                    Constraint::None => quote! {},
                    Constraint::Path(_) | Constraint::Ident(_) => quote! {(v)},
                };

                autoref_specializers.extend(quote! {
                    pub struct #tag;
                    impl #tag {
                        pub fn #into<T #t_constraint>(self, v: &T) -> #enum_name {
                            #enum_name::#variant #param
                        }
                    }

                    pub trait #r#match<T> {
                        fn #into_tag(&self) -> #tag;
                    }
                    impl<'t, T #t_constraint> #r#match<T> for #refs #wrap<'t,T> {
                        fn #into_tag(&self) -> #tag {
                            #tag
                        }
                    }
                });
            },
        )
        .count();

    autoref_specializers
}
