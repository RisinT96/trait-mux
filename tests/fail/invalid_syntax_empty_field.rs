use trait_mux::trait_mux;

fn main() {
    trait_mux!(,);
}
