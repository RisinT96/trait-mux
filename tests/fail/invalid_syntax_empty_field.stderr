error: expected identifier
 --> tests/fail/invalid_syntax_empty_field.rs:4:16
  |
4 |     trait_mux!(,);
  |                ^
